### ISSUE TITLE
Passive port ranges are not settable in t.p.ftp

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1904]
|<img alt="krp's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| krp reported|
|-|-|
|Trac ID|trac#1904|
|Type|enhancement|
|Created|2006-07-06 16:35:54Z|

As discussed on #twisted, FTP passive ports are chosed randomly by the OS (since we use reactor.listenTCP(0, f) for the dtp factory).

I propose this patch that offers the user the possibility to tune the passive port ranges.

To be reviewed, it may not fit twisted's style.

Attachments:

* [twisted-ftp-port-ranges.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a80/a80a6678fa430121ecc291bf4dcf2d27ab283d06/9246c8b37a806be29da1d7dfeb7c2145eb3a5d9c.patch) (1997 bytes) - added by krp on 2006-07-06 16:37:26Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1904 1904
type__enhancement enhancement
reporter__krp krp
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1152203754000000 1152203754000000
changetime__1188805821000000 1188805821000000
version__None None
owner__therve therve
cc__therve cc__spiv cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

This is ready to review in ftp-pasv-port-range-1904.

spiv, willing to have a look?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

So, here is a crazy thought

Port ranges should be determined by some method on the avatar (perhaps not a method which is part of the primary interface, though, since that seems to be oriented towards filesystem interaction, although maybe it'd be okay to add it there anyway).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_3):
> So, here is a crazy thought
>
> Port ranges should be determined by some method on the avatar (perhaps not a method which is part of the primary interface, though, since that seems to be oriented towards filesystem interaction, although maybe it'd be okay to add it there anyway).

It seems it's totally possible, but what does it bring exactly? We'll move the information of the ports from FTPFactory/FTP to FTPRealm/FTPShell. It looks relatively equivalent to me. If your point is for customisation, I've never seen a case where you want port settings customizable by client, but that could be useful for someone.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Hmm, you're right, it's a pretty rare use-case.  Okay, let's forget about that for now.

What I would like to see, then, is some simpler semantics for the range.  How about this, instead of having a min and a max (with 0 for min having a special meaning), there is a single attribute which gives a port range.  It can just be an iterator (eg, xrange(10, 20)).  getDtpPort can just iterate over it.  This should be less code and offer more flexibility.

How does this sound?

Incidentally, getDtpPort should capitalize DTP, since it's an acronym (Data Transfer Protocol).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [20940]) Use a range for passive port instead of 2 variables.

Refs [#1904](https://github.com/twisted/twisted/issues/1904)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

This sounds good :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Looking good, thanks!

I'd suggest some minor adjustments to the tests - I don't think any of them need to bind actual tcp ports.  Two of them are careful not to, but test_PASVportRange seems to want to, and I'm not sure why.  It could easily provide a listenFactory that asserts the beginning of the range is passed in first, and so on.  Also, it has a weird name :)  It might be okay to expand "pasv" to "passive" (maybe in the attribute name on FTP, too).

If test_PASVportRange stops using the real reactor.listenTCP, then it probably also won't have to take care to return the deferred returned by port.stopListening to avoid a cleanup error.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [21114]) 
 * Remove test using listenTCP
 * Rename pasvPortRange to passivePortRange

Refs [#1904](https://github.com/twisted/twisted/issues/1904).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

This test actually acts as a duplicate of test_portRange, so I removed it, and I rename the variable.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

Using list() on passivePortRange when raising CannotListenError from getDTPPort worries me a little.  I might not want to see a list of 5 thousand port numbers if that exception gets logged.  Maybe just having passivePortRange there would be okay?  xrange objects str kind of nicely on their own, at least.

test_portRangeForwardError still sets pasvMinPort which probably doesn't do anything anymore.

Some suggestions for docstrings:

  * test_portRangeForwardError - ```Exceptions other than L{error.CannotListenError} which are raised by C{listenFactory} should be raised to the caller of L{FTP.getDTPPort}.```
  * test_portRange - ```L{FTP.passivePortRange} should determine the ports which L{FTP.getDTPPort} attempts to bind.  If no port from that iterator can be bound, L{error.CannotListenError} should be raised, otherwise the first successful result from L{FTP.listenFactory} should be returned.```
  * FTP.listenFactory - ```@ivar listenFactory: A callable with the signature of L{twisted.internet.interfaces.IReactorTCP.listenTCP} which will be used to create Ports for passive connections (mainly for testing).```

These are all quite minor points.  I think you should be able to merge without another review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [21189]) Apply review comments.

Refs [#1904](https://github.com/twisted/twisted/issues/1904)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [21190]) Merge ftp-pasv-port-range-1904

Author: therve
Reviewer: exarkun
Fixes [#1904](https://github.com/twisted/twisted/issues/1904)

Throught the passivePortRange attribute of twisted.protocols.ftp.FTPFactory,
allow a FTP server to restrict the passive ports used instead of using a
random port.
